server/build.gradle

```diff
@@ -173,6 +173,17 @@ tasks.named("testingConventions").configure {
     }
 }
 
+// Read the list from maven central.
+// Fetch the metadata and parse the xml into Version instances, pick the latest one
+def latestRelease = new URL('https://repo1.maven.org/maven2/org/opensearch/opensearch/maven-metadata.xml').openStream().withStream { s ->
```

Comment:
This is clever - I like it.  However, if we need a breaking change for example for a security related feature.  Or better yet an accidentally applied `@PublicApi` tag, this check will fail until a new release has shipped, this could be many days or even weeks of acknowledging the same issue over and over again.

I have a counter proposal - what if we had different targets based on different branches?

### Leaf branches with an existing release such as 2.13 or unreleased 2.14
Use `latestRelease` as the target, getting annoyed by this check will be short-lived because we are about to release

### Major version branches such as 2.x or main
Use 'snapshotRelease' as the target, breaking changes will transit through them from time to time, but should not halt development.  These are often busier environments.


What would you think about having a parameter that switched between approaches?

Reply:
Thanks @peternied 

> However, if we need a breaking change for example for a security related feature. Or better yet an accidentally applied @PublicApi tag, this check will fail until a new release has shipped, this could be many days or even weeks of acknowledging the same issue over and over again.

Correct, the breaking change is unsolved problem (with this change, hence the GA check will still be not run) BUT I do have an idea how we could address that (this is also why the check is not enabled for `main` still).

> Use latestRelease as the target, getting annoyed by this check will be short-lived because we are about to release

So the primary target for this change is 2.x branch: where we have unreleased 2.14.0 is going to compared to released 2.13.0. 

> Use 'snapshotRelease' as the target, breaking changes will transit through them from time to time, but should not halt development. These are often busier environments.

`snapshotRelease`  is shaky target: for example, that could be new APIs which are only coming with new release but that could be changed / iterated over many times. While still in snapshot, the breaking changes will pass.

Reply:
> shaky target

Absolutely correct IMO this is a key reason to make sure we are exposing how to treat these failures well.  Such as comments or annotations on impacted lines of the PR, comments on the PR, or another nicer mechanism.

Note; if we declared something a Public API and then said whoops and made a breaking change I'd really like extra care in watching changes to make sure they have the correct design constraints in place

Reply:
>  Such as comments or annotations on impacted lines of the PR, comments on the PR, or another nicer mechanism.

Very right, but I think the code will not help here but the pull request itself (fe imaging whole classes are deleted, there is nothing to comment / annotate). The rough idea I have (which is not suggested anywhere yet) is to have in pull request description something like that:

``
/breaking <reason>
```

So when such pull request is built, the author acknowledge breaking change and provides the reason. This is just an idea at this moment.
 
> Note; if we declared something a Public API and then said whoops and made a breaking change I'd really like extra care in watching changes to make sure they have the correct design constraints in place

This switch happens only upon release, before that it is still just an API-in-work, right?

Reply:
> This switch happens only upon release, before that it is still just an API-in-work, right?

@reta I think we are splitting hairs, we both agree that breaking changes need a reconciliation process.  I think that this PR has captured an alternative detection process that is high value - checking against release version.

However, I think there is still value in running a detection check against recent snapshot, and so I'd like for both avenues to be available.  What do you think about making the target choice an argument to the gradle task with a default to your preference?

Reply:
@peternied done, thanks!